lib/cfg/cfg_Base.ml

```diff
@@ -0,0 +1,105 @@
+open Cfg_Node
+
+module CFG = struct
+  (* Auxiliary module for providing Hashtbl of a pretty printing function *)
+  module Hashtbl = struct
+    include Hashtbl
+
+    let pp pp_key pp_value ppf values =
+      Hashtbl.iter (fun key data -> Format.fprintf ppf "@[<1>%a: %a@]@." pp_key key pp_value data) values
+  end
+
+  type 'a item = {
+    idx   : int;
+    exp   : 'a;
+    pred  : int list;
+    succ  : int list;
+  } [@@deriving show]
+
+  (* The CFG is implemented as an Hashtable <id, (exp, predecessors, successors)> *)
+  type 'a t = {
+    cfg: (int, 'a item) Hashtbl.t;
+    root_id: int
+  }
+  [@@deriving show]
+
+  let make_item idx exp pred succ : 'a item =
+    {idx = idx; exp = exp; pred = pred; succ = succ}
+
+  (* starting from a node (the root), builds a CFG *)
+  let make (initial_node : 'a Node.t) : 'a t =
+    let h = Hashtbl.create (Node.length initial_node) in
+
+    let rec aux (h : (int, 'a item) Hashtbl.t) (node : 'a Node.t) : unit =
+      match Node.get_succ node with
+      | [] -> (* the node has not successors, insert it *)
+         Hashtbl.add h (Node.get_id node) (make_item (Node.get_id node) (Node.get_exp node) (Node.get_pred node) [])
```

Comment:
Could you please extract this in a separate function? To then use the new function also in the code below